contracts/pcv/PCVSwapperUniswap.sol

```diff
@@ -0,0 +1,314 @@
+
+import "./IPCVSwapper.sol";
+import "../refs/UniRef.sol";
+import "../utils/Timed.sol";
+import "../external/UniswapV2Library.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IWETH.sol";
+
+/// @title implementation for PCV Swapper that swaps ERC20 tokens on Uniswap
+/// @author eswak
+contract PCVSwapperUniswap is IPCVSwapper, UniRef, Timed {
+    using SafeERC20 for ERC20;
+    using Decimal for Decimal.D256;
+    using SafeMathCopy for uint256;
+
+    /// @notice the token to spend on swap (outbound)
+    address public tokenSpent;
+    /// @notice the token to receive on swap (inbound)
+    address public tokenReceived;
+    /// @notice the address that will receive the inbound tokens
+    address public tokenReceivingAddress;
+    /// @notice the maximum amount of tokens to buy (0 = unlimited)
+    uint256 public tokenBuyLimit = 0;
+    /// @notice the maximum amount of tokens to spend on every swap
+    uint256 public maxSpentPerSwap;
+    /// @notice should we use (1 / oraclePrice) instead of oraclePrice ?
+    bool public invertOraclePrice;
+    /// @notice the maximum amount of slippage vs oracle price
+    uint256 public maximumSlippageBasisPoints = 300; // default 3%
+    uint256 public constant BASIS_POINTS_GRANULARITY = 10_000;
+
+    constructor(
+        address _core,
+        address _pair,
+        address _router,
+        address _oracle,
+        uint256 _swapFrequency,
+        address _tokenSpent,
+        address _tokenReceived,
+        address _tokenReceivingAddress,
+        uint256 _maxSpentPerSwap,
+        bool _invertOraclePrice
+    ) public UniRef(_core, _pair, _router, _oracle) Timed(_swapFrequency) {
+        tokenSpent = _tokenSpent;
+        tokenReceived = _tokenReceived;
+        tokenReceivingAddress = _tokenReceivingAddress;
+        maxSpentPerSwap = _maxSpentPerSwap;
+        invertOraclePrice = _invertOraclePrice;
+
+        emit UpdateTokenSpent(_tokenSpent);
+        emit UpdateTokenReceived(_tokenReceived);
+
+        // start timer
+        _initTimed();
+    }
+
+    /// @notice All received ETH is wrapped to WETH
+    receive() external payable {
+      IWETH weth = IWETH(router.WETH());
+      weth.deposit{value: msg.value}();
+    }
+
+    // =======================================================================
+    // IPCVSwapper interface override
+    // =======================================================================
+
+    /// @notice withdraw ETH from the contract
+    /// @param to address to send ETH
+    /// @param amountOut amount of ETH to send
+    function withdrawETH(address payable to, uint256 amountOut) external override onlyPCVController {
+        Address.sendValue(to, amountOut);
+        emit WithdrawETH(msg.sender, to, amountOut);
+    }
```

Comment:
We'll need to unwrap WETH here before sending

Reply:
I think it's not possible to code this bevahior because when we call WETH.withdraw, the WETH contract sends ETH to the PCVSwapperUniswap, so the payable `receive()` function is called & wraps back to WETH... all my attempts to call `weth.withdraw(x)` revert. I think it makes sense to keep it in the interface though because not all PCVSwapper will wrap to WETH automatically.